### PR TITLE
CLN: avoid values_from_object in reshape.merge

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/reshape/merge.py [hunk 1]
```diff
@@ -1866,27 +1873,23 @@ def _factorize_keys(lk, rk, sort=True):
         lk = ensure_int64(lk.codes)
         rk = ensure_int64(rk)
 
-    elif (
-        is_extension_array_dtype(lk.dtype)
-        and is_extension_array_dtype(rk.dtype)
-        and lk.dtype == rk.dtype
-    ):
+    elif is_extension_array_dtype(lk.dtype) and lk.dtype == rk.dtype:
```
Comment (WillAyd) [inline pandas/core/reshape/merge.py]: Is this guaranteed to be equivalent?

**Review by WillAyd — COMMENTED** [review]
Can you check perf?

File: pandas/core/reshape/merge.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/reshape/merge.py]: I'm pretty sure.  How could you have this condition but not `is_extension_array_dtype(rk.dtype)`?

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -1849,9 +1850,15 @@ def _right_outer_join(x, y, max_groups):
...
 def _factorize_keys(lk, rk, sort=True):
     # Some pre-processing for non-ndarray lk / rk
+    lk = extract_array(lk, extract_numpy=True)
+    rk = extract_array(rk, extract_numpy=True)
+
     if is_datetime64tz_dtype(lk) and is_datetime64tz_dtype(rk):
-        lk = getattr(lk, "_values", lk)._data
-        rk = getattr(rk, "_values", rk)._data
+        # Extract the ndarray (UTC-localized) values
+        # Note: we dont need the dtypes to match, as these can still be compared
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/merge.py]: That does happen somewhere else?

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -1849,9 +1850,15 @@ def _right_outer_join(x, y, max_groups):
...
+        lk = lk._data
+        rk = rk._data
+        # TODO: this is equivalent to _values_for_factorize()[0]; more idiomatic?
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/merge.py]: I would do that yes

File: pandas/core/reshape/merge.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/reshape/merge.py]: yes, in the EA-dtype check below we require matching dtypes

File: pandas/core/reshape/merge.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/reshape/merge.py]: updated

File: pandas/core/reshape/merge.py (hunk 1 above)
Comment (jreback) [reply pandas/core/reshape/merge.py]: should change to lk.is_dtype_equal(rk) (technically the lk.dtype == rk.dtype is unsafe and we should never use this)

**Review by jreback — CHANGES_REQUESTED** [review]
i am thinking this actually enables some cases? though we do have pretty comprehensive testing here.

comment

**Comment by jbrockmendel** [comment]
> i am thinking this actually enables some cases? 

not sure i follow.  can you give an example?

File: pandas/core/reshape/merge.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/reshape/merge.py]: will update

**Comment by jreback** [comment]
> > i am thinking this actually enables some cases?
> 
> not sure i follow. can you give an example?

are there any merge cases that work that didn't previously?

**Comment by jbrockmendel** [comment]
> are there any merge cases that work that didn't previously?

there are some dtype combinations that arent tested that looked sketchy.  there's a TODO comment about adding tests for them

**Comment by jreback** [comment]
> > are there any merge cases that work that didn't previously?
> 
> there are some dtype combinations that arent tested that looked sketchy. there's a TODO comment about adding tests for them

kk

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. merge o ngreen.

**Comment by jreback** [comment]
hmm, something broke

**Comment by jbrockmendel** [comment]
updated+green

**Comment by WillAyd** [comment]
Thanks @jbrockmendel

**Pull request merged**